### ISSUE TITLE
hyperopt can not successfully run in driver using a remote Ray Cluster

### ISSUE THREAD

**Issue by Jeffwan** [ludwig-ai/ludwig#2401]
**Describe the bug**

I am playing the Ludwig in distributed environment and meet some issues. I can not run hyperopt successfully in a remote Ray Cluster. Seems that's a ray tune issue and not sure if Ludwig community can help on that. 

**To Reproduce**

1. Create a ray cluster on kubernetes. I expose nodePort for external clients to connect to RAY HEAD.
2. Create a driver pod like below and it will talk to the remote Ray cluster using address I specified. 

```
apiVersion: batch/v1
kind: Job
metadata:
  name: x-automl-driver
  namespace: ray-system
spec:
  template:
    spec:
      containers:
      - name: ludwig
        image: seedjeffwan/automl:v0.3.3
        command: ["sleep", "3600"]
        env:
        - name: RAY_ADDRESS
          value: ray://35.80.xxx.xxx:30348
        - name: AWS_ACCESS_KEY_ID
          value: xxxx
        - name: AWS_SECRET_ACCESS_KEY
          value: xxxx
      restartPolicy: Never
```

I enter the pod to run 

```
ludwig hyperopt --config s3://ludwig-automl/rotten_tomatoes.yaml --dataset s3://ludwig-automl/rotten_tomatoes.csv
```

```
ludwig hyperopt --config s3://ludwig-automl/rotten_tomatoes.yaml --dataset s3://ludwig-automl/rotten_tomatoes.csv
NumExpr defaulting to 4 threads.
/usr/local/lib/python3.8/site-packages/sklearn/experimental/enable_hist_gradient_boosting.py:16: UserWarning: Since version 1.0, it is not needed to import enable_hist_gradient_boosting anymore. HistGradientBoostingClassifier and HistGradientBoostingRegressor are now stable and can be normally imported from sklearn.ensemble.
  warnings.warn(
Found credentials in environment variables.
Using address ray://35.80.xxx.xxx:30348 set in the environment variable RAY_ADDRESS
Passing the following kwargs to ray.init() on the server: ignore_reinit_error
Ray processor params: None
███████████████████████
█ █ █ █  ▜█ █ █ █ █   █
█ █ █ █ █ █ █ █ █ █ ███
█ █   █ █ █ █ █ █ █ ▌ █
█ █████ █ █ █ █ █ █ █ █
█     █  ▟█     █ █   █
███████████████████████
ludwig v0.6.dev - Hyperopt

Hyperopt config
{   'executor': {   'cpu_resources_per_trial': 1,
                    'num_samples': 10,
                    'scheduler': {   'grace_period': 72,
                                     'max_t': 300,
                                     'reduction_factor': 5,
                                     'time_attr': 'time_total_s',
                                     'type': 'async_hyperband'},
                    'time_budget_s': 300,
                    'type': 'ray'},
    'goal': 'maximize',
    'metric': 'roc_auc',
    'output_feature': 'recommended',
    'parameters': {   'combiner.bn_momentum': {   'categories': [   0.4,
                                                                    0.3,
                                                                    0.2,
                                                                    0.1,
                                                                    0.05,
                                                                    0.02],
                                                  'space': 'choice'},
                      'combiner.bn_virtual_bs': {   'categories': [   256,
                                                                      512,
                                                                      1024,
                                                                      2048,
                                                                      4096],
                                                    'space': 'choice'},
                      'combiner.num_steps': {   'categories': [   3,
                                                                  4,
                                                                  5,
                                                                  6,
                                                                  7,
                                                                  8,
                                                                  9,
                                                                  10],
                                                'space': 'choice'},
                      'combiner.output_size': {   'categories': [   8,
                                                                    16,
                                                                    24,
                                                                    32,
                                                                    64,
                                                                    128],
                                                  'space': 'choice'},
                      'combiner.relaxation_factor': {   'categories': [   1.0,
                                                                          1.2,
                                                                          1.5,
                                                                          4.0],
                                                        'space': 'choice'},
                      'combiner.size': {   'categories': [8, 16, 24, 32, 64],
                                           'space': 'choice'},
                      'combiner.sparsity': {   'categories': [   0.0,
                                                                 1e-06,
                                                                 0.0001,
                                                                 0.001,
                                                                 0.01,
                                                                 0.1],
                                               'space': 'choice'},
                      'trainer.decay_rate': {   'categories': [0.8, 0.9, 0.95],
                                                'space': 'choice'},
                      'trainer.decay_steps': {   'categories': [   500,
                                                                   2000,
                                                                   8000,
                                                                   10000,
                                                                   20000],
                                                 'space': 'choice'},
                      'trainer.learning_rate': {   'categories': [   0.005,
                                                                     0.01,
                                                                     0.02,
                                                                     0.025],
                                                   'space': 'choice'}},
    'search_alg': {'random_state_seed': None, 'type': 'hyperopt'},
    'split': 'validation'}


Features that may be updated in hyperopt trials if default parameters are specified in the search space
{   'input_features': defaultdict(<class 'set'>,
                                  {   'binary': {'top_critic'},
                                      'category': {   'content_rating',
                                                      'genres',
                                                      'movie_title'},
                                      'number': {'runtime'}}),
    'output_features': defaultdict(<class 'set'>, {'binary': {'recommended'}})}


Using full raw dataset, no hdf5 and json file with the same name have been found
Building dataset (it may take a while)
[########################################] | 100% Completed |  0.2s
(dask:('to-parquet-d14fd399da13c4a97e815626d621126a', 0) pid=728) Found credentials in environment variables.
Building dataset: DONE
[########################################] | 100% Completed |  0.1s
[########################################] | 100% Completed |  0.1s
[########################################] | 100% Completed |  0.1s
2022-08-19 18:18:04,624	WARNING read_api.py:252 -- The number of blocks in this dataset (1) limits its parallelism to 1 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
Caught schedule exception
2022-08-19 18:18:04,760	WARNING read_api.py:252 -- The number of blocks in this dataset (1) limits its parallelism to 1 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
2022-08-19 18:18:04,847	WARNING read_api.py:252 -- The number of blocks in this dataset (1) limits its parallelism to 1 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
2022-08-19 18:18:04,858	INFO common.py:219 -- Exception from actor creation is ignored in destructor. To receive this exception in application code, call a method on the actor reference before its destructor is run.
(run pid=820) /usr/local/lib/python3.8/site-packages/sklearn/experimental/enable_hist_gradient_boosting.py:16: UserWarning: Since version 1.0, it is not needed to import enable_hist_gradient_boosting anymore. HistGradientBoostingClassifier and HistGradientBoostingRegressor are now stable and can be normally imported from sklearn.ensemble.
(run pid=820)   warnings.warn(
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/ludwig/hyperopt/execution.py", line 794, in execute
    analysis = tune.run(
  File "/usr/local/lib/python3.8/site-packages/ray/tune/tune.py", line 413, in run
    return ray.get(remote_future)
  File "/usr/local/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 104, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/api.py", line 44, in get
    return self.worker.get(vals, timeout=timeout)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/worker.py", line 438, in get
    res = self._get(to_get, op_timeout)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/worker.py", line 466, in _get
    raise err
types.RayTaskError(AttributeError): ray::run() (pid=820, ip=192.168.37.141)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/tune.py", line 479, in run
    is_function_trainable(trainable)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 60, in is_function_trainable
    trainable = get_trainable_cls(trainable)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 44, in get_trainable_cls
    return _global_registry.get(TRAINABLE_CLASS, trainable_name)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 196, in get
    return pickle.loads(value)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/common.py", line 100, in __init__
    self._set_id(id)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/common.py", line 191, in _set_id
    self._worker.call_retain(id)
AttributeError: 'NoneType' object has no attribute 'call_retain'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/ludwig", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/site-packages/ludwig/cli.py", line 166, in main
    CLI()
  File "/usr/local/lib/python3.8/site-packages/ludwig/cli.py", line 66, in __init__
    getattr(self, args.command)()
  File "/usr/local/lib/python3.8/site-packages/ludwig/cli.py", line 91, in hyperopt
    hyperopt_cli.cli(sys.argv[2:])
  File "/usr/local/lib/python3.8/site-packages/ludwig/hyperopt_cli.py", line 409, in cli
    hyperopt_cli(**vars(args))
  File "/usr/local/lib/python3.8/site-packages/ludwig/hyperopt_cli.py", line 161, in hyperopt_cli
    return hyperopt(
  File "/usr/local/lib/python3.8/site-packages/ludwig/hyperopt/run.py", line 338, in hyperopt
    hyperopt_results = hyperopt_executor.execute(
  File "/usr/local/lib/python3.8/site-packages/ludwig/hyperopt/execution.py", line 823, in execute
    raise RuntimeError(f"Encountered Ray Tune error: {e}")
RuntimeError: Encountered Ray Tune error: ray::run() (pid=820, ip=192.168.37.141)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/tune.py", line 479, in run
    is_function_trainable(trainable)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 60, in is_function_trainable
    trainable = get_trainable_cls(trainable)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 44, in get_trainable_cls
    return _global_registry.get(TRAINABLE_CLASS, trainable_name)
  File "/usr/local/lib/python3.8/site-packages/ray/tune/registry.py", line 196, in get
    return pickle.loads(value)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/common.py", line 100, in __init__
    self._set_id(id)
  File "/usr/local/lib/python3.8/site-packages/ray/util/client/common.py", line 191, in _set_id
    self._worker.call_retain(id)
AttributeError: 'NoneType' object has no attribute 'call_retain'
```

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**

- Python version 3.8
- Ludwig version - 0.6.0-dev nightly image

**Additional context**

**Comment by Jeffwan** [comment]
The interesting thing is if I run the exact same command inside the head pod. It works.  I am wondering if there's connection issue from driver to ray rune?  

This is exact https://github.com/ludwig-ai/ludwig/issues/2153 wants to figure out. It would be great to add some guidance to run ludwig in distributed environment.

**Comment by ShreyaR** [comment]
Thanks for filing this issue @Jeffwan! We're currently looking into this to see if we can repro the issue and will post more updates soon.

**Comment by ShreyaR** [comment]
@Jeffwan this issue was caused because we register the trainable function used in `tune.run`, and retrieve the registered trainable on the worker nodes. This caused an issue because `register_trainable` uses the KV store in Ray, which is probably not available when using Ray client.

I've updated the code to account for this, and a fix is in progress in #2501. This fix will be included in the upcoming 0.6 Ludwig release.
